charmhelpers/contrib/charmsupport/nrpe.py

```diff
@@ -313,8 +321,24 @@ def write(self):
         monitor_ids = relation_ids("local-monitors") + \
             relation_ids("nrpe-external-master")
         for rid in monitor_ids:
+            reldata = relation_get(unit=local_unit(), rid=rid)
+            if 'monitors' in reldata:
+                old_monitors = yaml.safe_load(reldata['monitors'])
+                old_nrpe_monitors = old_monitors['monitors']['remote']['nrpe']
+                for shortname in old_nrpe_monitors.copy():
+                    if shortname in self.remove_check_queue:
+                        del old_nrpe_monitors[shortname]
+                    elif shortname in nrpe_monitors:
+                        old_nrpe_monitors[shortname] = nrpe_monitors[shortname].copy()
+                        del nrpe_monitors[shortname]
+                for shortname in nrpe_monitors:
+                    old_nrpe_monitors[shortname] = nrpe_monitors[shortname]
+                monitors = old_monitors.copy()
```

Comment:
There is no need to make a second copy of the dict, because yaml.dump isn't going to mutate it.

Reply:
I'm a bit confused by this code.  Reviewing alongside the existing code, here's what I *think* it is doing.

1. The existing code overwrote the relation data with newly calculated sets of checks.
2. The idea is to preserve existing entries unless they have been explicitly removed.
3. Any existing (in the relation data) monitor entries should be updated with new information (based on the newly computed info).

If that is the case, then the above code (for me) obfuscates it; thus it might become a maintenance issue.

The bit that seems to be changing is `nrpe_monitors` dictionary and the keynames appear to be the bit that is checked.  Essentially, remove elements that are in `self.remove_check_queue` and update any existing elements from the new `monitors` dict.

Curiously, only update the first occurrence of shortname found in `nrpe_monitors` as it is then deleted so it can't be used again.  Is that what's really wanted?  I'm not sure if that is a bug or not?

Anyway, I'm wondering if this is the code that you want:

```python
for rid in monitor_ids:
    reldata = relation_get(unit=local_unit(), rid=rid)
    if 'monitors' in reldata:
        # update the existing set of monitors with the new data
        old_monitors = yaml.safe_load(reldata['monitors'])
        old_nrpe_monitors = old_monitors['monitors']['remote']['nrpe']
        # remove keys that are in the remove_check_queue
        old_nrpe_monitors = {k: v for k, v in old_nrpe_monitors.items() if k not in self.remove_check_queue}
        # update/add nrpe_monitors
        old_nrpe_monitors.update(nrpe_monitors)
        # OPTIONAL (if needed): remove remove nrpe_monitor shortnames from nrpe_monitors (i.e. they only get applied once???)
        nrpe_monitors = {k: v for k, v in nrpe_monitors.items() if k in old_nrpe_monitors}
        # write back to the relation
        relation_set(relation_id=rid, monitors=yaml.dump(old_monitors))
    else:
        # write a brand new set of monitors, as no existing ones.
        relation_set(relation_id=rid, monitor=yaml.dump(monitors))
```